File: src/main/java/my/study/repository/design/pattern/decorator/java/DecorateA.java
package my.study.repository.design.pattern.decorator.java;

/**
 * @author djh on  2018/10/2 20:35
 * @E-Mail 1544579459@qq.com
 */
class DecorateA extends BaseDecorate {

    DecorateA(BaseComponent baseComponent) {
        super(baseComponent);
    }

    @Override
    void operator() {
        decorateMethod();
        super.operator();
    }

    private void decorateMethod() {
        System.out.println("装饰行为!");
    }
}

File: src/main/java/my/study/repository/design/pattern/decorator/kotlin/DecorateA.kt
package my.study.repository.design.pattern.decorator.kotlin

/**
 * @author djh on  2018/10/2 20:41
 * @E-Mail 1544579459@qq.com
 */
class DecorateA
(private val concreteComponent: ConcreteComponent = ConcreteComponent())
    : BaseComponent by concreteComponent {

    override fun operator() {
        concreteComponent.operator()
        decorateMethod()
    }

    private fun decorateMethod() {
        println("装饰方法!")
    }
}